### DISCUSSION TITLE
Store full PC2 in Rtabmap Database

### DISCUSSION THREAD

**Discussion by Nosille** [introlab/rtabmap#1428] [Ideas]
While I understand that there are significant computational and storage efficiencies associated with the built-in pcl datatypes (e.g., pcl::PointCloud<pcl::PointXYZRGBNormal>), I find these datatypes very limiting.  I am a ROS user, and I frequently find myself wishing RTABMAP would simply store the full original PointCloud2 object in the database.  As a result, I have been working on a modified version of RTABMAP that adds three new database entries, pc2_info, pc2_fields, and pc2_data.  These entries are simply a breakdown of a PCLPointCloud2 objects into blobs that can be stored in a  sql database.  I am close to having a usable product and would love the official RTABMAP team's advice and feedback moving forward.  My hope is that this approach will provide a lot of flexibility for data storage.  It many cases, I believe it will provide a very straight forward path for capturing user generated data.

https://github.com/Nosille/rtabmap/tree/AddPc2ToDatabase
https://github.com/Nosille/rtabmap_ros/tree/AddPc2ToDatabase

I am currently creating the scan and scan_info blobs consistent with the main branch and simultaneously populating the pc2_info, pc2_fields, and pc2_data blobs with the full original PCLPointCloud2 object. While this means that basic pointcloud information is stored twice right now, I believe the efficiency of pcl::PointCloud objects for common operations justifies keeping the scan/scan_info blobs, and I wasn't ready to couple the scan and pc2 database entries in a way that would allow the pc2 data to exist without the x, y, z fields.  Hopefully, it is something that can be improved in the future.

**Comment by matlabbe** [comment]
You may have succeeded to do something similar by extending formats supported by LaserScan object (for the format missing of your PointCloud2):
https://github.com/introlab/rtabmap/blob/f8b1d5065618b481cccfd777bb086744c0440962/corelib/include/rtabmap/core/LaserScan.h#L40-L51 

then create a new global map object similar to [CloudMap](https://github.com/introlab/rtabmap/blob/master/corelib/src/global_map/CloudMap.cpp) (assembling [LocalGrid](https://github.com/introlab/rtabmap/blob/master/corelib/src/LocalGrid.cpp)) that could be named like `LaserScanCloudMap` using [LaserScan](https://github.com/introlab/rtabmap/blob/master/corelib/src/LaserScan.cpp) member of each SensorData, in order to assemble the output cloud on ROS side (in same format than your input PointCloud2 format).

For `PointXYZRGBNormal` format, are you referring to [ExportCloudsDialog](https://github.com/introlab/rtabmap/blob/master/guilib/src/ExportCloudsDialog.cpp)? I agree that we cannot export the laser scans in exact format than the one in LaserScan object (even if we add for compatible formats), that class would need to be refactored. Maybe by reusing the new `LaserScanCloudMap` class. `ExportCloudsDialog` supports many options (with/without normals, with/without colored RGB points or intensity, ...), which make difficult to use PCLPointCloud2. Note that in case of lidar, `PointXYZRGBNormal` is converted to `PointXYZINormal` before saving, so we can recover Intensity in the output file.
https://github.com/introlab/rtabmap/blob/f8b1d5065618b481cccfd777bb086744c0440962/guilib/src/ExportCloudsDialog.cpp#L4298-L4303

Another thing is that over the years, `LaserScan` looks more and more to a `PCLPointCloud2::Ptr` object (which was not the case at the beginning where only 2D scans were supported). Changing `cv::Mat` format used inside `LaserScan` to `PCLPointCloud2::Ptr` format could be a more sustainable solution, though a large refactoring task.

**Reply by Nosille** [reply]
I considered extending or modifying LaserScan in a manner similar to you suggest, but it was simply too intimidating a refactor.  I did not want to break the entire program and have to piece it back together in a single major effort.  This dual database entry approach has allowing me to approach each module/function as I find time.  Along those lines, I recently managed to get the ExportCloudDialog to work with the new database entries.  I also created new LocalCloud and LocalCloudMaker classes as PointCloud2 variants of LocalGrid and LocalGridMaker.  These new classes let me pass the data back to ROS.  As you noted, both were major refactors of the underlying code.  I ended up switching the final working data in all 3 functions to a PCLPointcloud2, using your existing functions with a conversion to pointcloud2, and adding functions for the PC2 database entries based on a user setting.

https://github.com/Nosille/rtabmap/tree/ExportPc2Cloud
https://github.com/Nosille/rtabmap_ros/tree/Pc2InRviz
